Mi23307

perisicMarko opened the pull request:
FragPos  tj aPos argument svakog fragmenta mi je iz nekog razloga 0, 0, 0,

perisicMarko committed:
Build plain app
Set app to close on esc

perisicMarko committed:
rename the window title
add skybox

perisicMarko committed:
change skybox
add moon model

perisicMarko committed:
add wasd(keyboard) camera navigation
make observer for mouse change
add depth testing
set cursor hidden

perisicMarko committed:
add space_station model
change window title to Space station
set first mouse flick skipped

perisicMarko committed:
add space_craft model
change window size
adjust size and position of all added models

perisicMarko committed:
add pointer light from moon source to ship and station

perisicMarko committed:
update PROJECT-DESCRIPTION-TEMPLATE.md
add TODO list

perisicMarko committed:
add gui with customizable elements
update TODO.md and PROJECT-DESCRIPTION-TEMPLATE.md

perisicMarko committed:
Svaki fragment position vertex arguemnt mi je nula, u bilo kom modelu
stavio sam da se modeli boje u crveno ako im je fpos = 0

spaske00 commented on app/CMakeLists.txt:
```diff
@@ -1,11 +1,15 @@
 project(APP)
 file(GLOB sources src/*.cpp)
-file(GLOB headers include/*.hpp)
+file(GLOB headers include/*.h)
```
Vratiti na .hpp i preimenovati sva zaglavlja na .hpp jer je izvorni kod C++.
Zaglavlje .h se koristi samo u slučaju da je izvorni kod u jeziku C.
CMake fajlove ne bi trebalo menjati jer već rade sve što je potrebno da bi se projekat kompilirao.

spaske00 commented on app/CMakeLists.txt:
```diff
@@ -1,11 +1,15 @@
+        src/MyApp.cpp
+        include/MyApp.h
+        src/MainController.cpp
+        include/MainController.h)
```
Slobodno obrisati sve četiri linije, cmake će automatski pokupti ove fajlove preko ${sources} i ${headers}

spaske00 commented on app/src/Main.cpp:
```diff
@@ -1,8 +1,9 @@
-#include <engine/core/Engine.hpp>
+#include <iostream>
```
Nepotrebno zaglavlje.

spaske00 commented on app/src/MainController.cpp:
```diff
@@ -0,0 +1,236 @@
+#include <engine/graphics/OpenGL.hpp>
+#include <engine/platform/PlatformController.hpp>
+#include <engine/resources/ResourcesController.hpp>
+#include "../include/MainController.h"
```
Dovoljno je `#include <MainController.hpp>

perisicMarko replied:
Mislite da prebacim u .hpp pa je dovoljno?

spaske00 replied:
Oprostite, nisam se najpreciznije izrazio. 
Dakle linija `include_directories(include/)` https://github.com/marko11120/matf-rg-project-2024/blob/848c859d68e161d8eb73ebc905348f6930f79ab7/app/CMakeLists.txt#L3-L6 

zapravo stavlja kompajleru na putanju za zaglavlja ceo `include` direktorijum, tako da je u .cpp fajlovima dovoljno samo navesti njihovo ime, nema potrebe za relativnom putanjom. 

Kada `MainController.h` bude preimenovan u `MainController.hpp` umesto 
`#include "../include/MainController.h"`
biće dovoljno
`#include <MainController.hpp>`

spaske00 commented on app/src/MainController.cpp:
```diff
@@ -0,0 +1,236 @@
+#include <engine/resources/ResourcesController.hpp>
+#include "../include/MainController.h"
+
+#include <GLFW/glfw3.h>
```
Uh, nikako zaglavlja i biblioteke specifične za implementaciju u modul `app`. 
Projekat se neće ni kompilirati ako je ovo zaglavlje uključeno van `Engine`.

spaske00 commented on app/src/MainController.cpp:
```diff
@@ -0,0 +1,236 @@
+
+#include <GLFW/glfw3.h>
+#include <imgui.h>
+#include <iostream>
```
Koristite spdlog za logovanje. 
Ovo zaglavlje isto možete obrisati.

spaske00 commented on app/src/MainController.cpp:
```diff
@@ -0,0 +1,236 @@
+#include <imgui.h>
+#include <iostream>
+
+#include "glm/gtc/type_ptr.hpp"
```
Koristiti <> za biblioteke, i generalno za sva ostala uključivanja zaglavlja.

spaske00 commented on app/src/MainController.cpp:
```diff
@@ -0,0 +1,236 @@
+class MainPlatformEventObserver : public engine::platform::PlatformEventObserver {
+    void on_mouse_move(engine::platform::MousePosition position) override;
+    void on_key(engine::platform::Key key) override;
+    bool first_flick = true;
```
Ispratite konveciju za imenovanje promenljivih iz ostatka projekta.
Private polja uvek imaju prefiks `m_`, dakle `m_first_flick`.

spaske00 commented on app/src/MainController.cpp:
```diff
@@ -0,0 +1,236 @@
+    if(platform->key(engine::platform::KeyId::KEY_F12).is_down()) {
+        if(!cursor_visibility) {
+            cursor_visibility = true;
+            camera->set_cursor_status(cursor_visibility);
```
Hm, nisam siguran da je potreban cursor status na klasi `Camera`. Proveru stanja kursora bi trebalo da bude u `PlatformController`-u, odakle mu onda možete pristupiti iz celog projekta.

spaske00 commented on app/src/MainController.cpp:
```diff
@@ -0,0 +1,236 @@
+            cursor_visibility = false;
+            camera->set_cursor_status(cursor_visibility);
+            camera->set_cursor_visible(cursor_visibility);
+        }
```
Deluje da je ceo if zapravo:
```
cursor_visibility = !cursor_visiblity;
camera->set_cursor_status(cursor_visibility);
camera->set_cursor_visible(cursor_visibility);
```

spaske00 commented on app/src/MainController.cpp:
```diff
@@ -0,0 +1,236 @@
+    shader->set_mat4("projection", graphics->projection_matrix());
+    shader->set_mat4("view", graphics->camera()->view_matrix());
+    glm::mat4 model_matrix = glm::mat4(1.0f);
+    float timeValue = 10 * glfwGetTime();
```
Neće se kompilirati, glfw nije dostupna iz `app`.
Mislim da je ideja ovde: `PlatformController::frame_time().current`

spaske00 commented on app/src/MainController.cpp:
```diff
@@ -0,0 +1,236 @@
+    auto resources = engine::core::Controller::get<engine::resources::ResourcesController>();
+    auto graphics = engine::core::Controller::get<engine::graphics::GraphicsController>();
+    engine::resources::Model *model   = resources->model("space_station");
+    engine::resources::Shader *shader = resources->shader("space_station");
```
`space_station` shader ne postoji u projektu

spaske00 commented on app/src/MainController.cpp:
```diff
@@ -0,0 +1,236 @@
+    auto resources = engine::core::Controller::get<engine::resources::ResourcesController>();
+    auto graphics = engine::core::Controller::get<engine::graphics::GraphicsController>();
+    engine::resources::Model *model   = resources->model("spacecraft2");
+    engine::resources::Shader *shader = resources->shader("space_craft");
```
space_craft shader ne postoji u projektu

spaske00 commented on engine/include/engine/graphics/Camera.hpp:
```diff
@@ -113,11 +113,15 @@ namespace engine::graphics {
          */
         void update_camera_vectors();
+        bool cursor_visible;
     };
```
Mislim da ovo definitivno ne treba da bude u klasi Camera, kursor i stanje kursora je odgovornost platforme, tako da bi najbolje bilo premestiti postavljanje i proveru stanja u `PlatformController`.

spaske00 reviewed (changes requested):
Molim Vas, ispravite prvo komentare u kodu i javite mi kada završite da pogledam ponovo :)

perisicMarko committed:
post pull request code refactoring
-cursor stats from Camera.hpp to PlatformController.hpp
-minor code cleansing
-private class fields renamed to m_ + field_name

perisicMarko commented:
Pitanje vezano za github, sad sam pregazio kod sa novim push-om ali ostaje los commit u istoriji. Da li je praksa da na lokalnom rep-u resetujem na prethodni koji je validan pa ponovo da push-am sa force flag-om?. Sada sam sam na projektu pa je svejedno, ali da je vise ljudi na projektu neko je mogao da povuce kod sa tim commit-om i kada ja push-am sa force-om sve bez tog commit-a, sta ce se desiti? Kapiram da ce se razlikovati commit log ali kako se ta situacija resava.

spaske00 commented:
> Pitanje vezano za github, sad sam pregazio kod sa novim push-om ali ostaje los commit u istoriji.

Koliko mogu da vidim samo ste pushovali novi komit na grani koji sadrži ispravke na osnovu mojih komentara. To je skroz ok. 

> Da li je praksa da na lokalnom rep-u resetujem na prethodni koji je validan pa ponovo da push-am sa force flag-om?  

Nema potrebe, dobro je da komitovi ostanu na grani **dok je u razvoju** kako bi se videle sve promene.  

> Sada sam sam na projektu pa je svejedno, ali da je vise ljudi na projektu neko je mogao da povuce kod sa tim commit-om i kada ja push-am sa force-om sve bez tog commit-a, sta ce se desiti?

Na stranici kursa u sekciji sa dokumentacijom možete naći link ka interaktivnom okruženju za učenje kako git radi u pozadini. Možete tamo probati ovaj eksperiment i saznati :) 

> Kapiram da ce se razlikovati commit log ali kako se ta situacija resava. 

Kada radite na nekom delu projekta napravite novu granu. Sve promene postavljate samo na tu granu. Kada je posao završen, komitove možete spljoštiti (squash) i zatim spojiti (merge) Vašu granu sa glavnom granom. 
Ukoliko su se na glavnoj grani desile promene od kojih zavisi i Vaša deo posla, možete te promene preuzeti sa `rebase` ili `merge`.  

Molim Vas, za dalje lekcije, ispratite upustvo o radu sa granama na stranici kursa u delu `projekat`, tamo je postupak detaljno opisan i javite naravno ako negde zapne :)

perisicMarko commented:
Ostao sam na istom problemu, kao i pre ispravki koje ste sugerisali, i dalje iz nekog razloga ako posaljem tacku (1,1,1) out FragPos argumentu u fragment shader  tamo mi taj FragPos bude (0,0,0). U kodu sam ostavio sa aPos pomnozen model matricom. Ista situacija je i u space_station.glsl i space_craft.glsl.

spaske00 commented:
Molim Vas okačite i sliku ekrana sa modelima. 
Evo još par stvari koji možete uraditi:
1. Prikažite modele u engine/test/app, ako tako rade, znači greška je u Vašem delu koda, ne u modelu. 
2. Proverite da li su vrteksi ispravno učitani: Breakpoint u ResourcesController.cpp na učitavanju Assimp scene i onda proverite da li su učitani vrteksi različiti od nule.
3.. Pogeldajte alat `renderdoc` iz dokumentaciji, pomoću njega možete uhvatiti svaki frajem i proveriti vrednosti svih vrteksa.

U kom komitu se tačno taj bag pojavio i kako ste ga oktrili? Vidim u framgnet šejderu neke provere da je FragPos.x < 0.1, ali to ne mora značiti da su pozicijie nula vektori.

perisicMarko commented:
Resio sam problem, tacnije ja sam mislio da ga imam jer mi je ulazio u taj if gde proveravam da li je argument nula. Verovatno sam ispravio deo koda koji je bio neispravan nakon sto sam stavio taj if kao proveru i nisam ni znao da sam ispravio jer mi je ulazio u taj if. Hvala :)

Jedno tehnicko pitanje obzirom da je vanredna situacija i da je datum na sajtu za konsultacije oko projekta probijen. Do kog datuma smemo da se javljamo za konsultacije i procenu projekta?

spaske00 commented:
> Resio sam problem, tacnije ja sam mislio da ga imam jer mi je ulazio u taj if gde proveravam da li je argument nula. Verovatno sam ispravio deo koda koji je bio neispravan nakon sto sam stavio taj if kao proveru i nisam ni znao da sam ispravio jer mi je ulazio u taj if. Hvala :)

To je to, taj if mi je baš bio čudan. :) 

> Jedno tehnicko pitanje obzirom da je vanredna situacija i da je datum na sajtu za konsultacije oko projekta probijen. Do kog datuma smemo da se javljamo za konsultacije i procenu projekta?

Dokle god ne bude objavljen formular za prijavu možete slati pitanja za projekte. Kada se tačni datumi ispita budu znali blagovremeno će i informacije na stranici kursa biti objavljene. Naravno, uzeće se u obzir da je situacija nije uobičajna, pa će i datumi i rokovi biti korigovani shodno tome.

perisicMarko commented:
Jasno, hvala.
